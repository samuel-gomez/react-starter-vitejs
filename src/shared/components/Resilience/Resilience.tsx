import { ComponentProps, PropsWithChildren, ReactNode } from 'react';
import ResilienceSubstitut from './ResilienceSubstitut';

type TResilience<Trefetch> = Omit<ComponentProps<typeof ResilienceSubstitut>, 'anomaly'> & {
  anomaly?: ComponentProps<typeof ResilienceSubstitut>['anomaly'] | null;
  refetch?: Trefetch;
  children?: ReactNode;
};

const Resilience = <Trefetch extends React.MouseEventHandler<HTMLButtonElement>>({
  resilienceMode,
  resilienceModifier,
  FallbackComponent,
  refetch,
  anomaly = null,
  children = null,
}: PropsWithChildren<TResilience<Trefetch>>) =>
  anomaly ? (
    <ResilienceSubstitut<Trefetch>
      anomaly={anomaly}
      refetch={refetch}
      resilienceMode={resilienceMode}
      resilienceModifier={resilienceModifier}
      FallbackComponent={FallbackComponent}
    />
  ) : (
    <>{children}</>
  );

export default Resilience;
