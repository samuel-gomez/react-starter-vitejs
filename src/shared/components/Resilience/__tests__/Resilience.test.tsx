import { describe, it, expect } from 'vitest';
import { render } from '@testing-library/react';
import Resilience from '../Resilience';

describe('Resilience', () => {
  const CustomComponent = () => <div>No anomaly</div>;

  const resilienceProps = {
    resilienceModifier: '',
  };

  it('Should render null when Resilience have been called with no anomaly and no child', () => {
    const EnhancedComponent = () => <Resilience {...resilienceProps} />;
    const { asFragment } = render(<EnhancedComponent />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('Should render child Component when Resilience have been called with no anomaly and child', () => {
    const EnhancedComponent = () => (
      <Resilience {...resilienceProps}>
        <CustomComponent />
      </Resilience>
    );
    const { getByText } = render(<EnhancedComponent />);
    expect(getByText('No anomaly')).toBeDefined();
  });

  it('Should render FallbackComponent when Resilience have been called with anomaly and FallbackComponent', () => {
    const ErrorComponent = () => <div>There is an error</div>;
    const anomaly = {
      label: 'error label',
    };

    const EnhancedComponent = () => (
      <Resilience {...resilienceProps} resilienceMode="fallback" FallbackComponent={ErrorComponent} anomaly={anomaly}>
        <CustomComponent />
      </Resilience>
    );
    const { getByText } = render(<EnhancedComponent />);
    expect(getByText('There is an error')).toBeDefined();
  });
});
