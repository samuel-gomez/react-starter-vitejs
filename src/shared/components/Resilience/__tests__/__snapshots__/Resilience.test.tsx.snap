// Vitest Snapshot v1

exports[`Resilience > Should render null when Resilience have been called with no anomaly and no child 1`] = `<DocumentFragment />`;
